alf/algorithms/algorithm.py

```diff
@@ -1173,41 +1284,87 @@ def train_from_replay_buffer(self, update_global_counter=False):
+            # prioritized replay, then make sure replay and train below is atomic.
+            with record_time("time/replay"):
+                mini_batch_size = config.mini_batch_size
+                if mini_batch_size is None:
+                    mini_batch_size = self._replay_buffer.num_environments
+                if config.whole_replay_buffer_training:
+                    experience, batch_info = self._replay_buffer.gather_all(
+                        ignore_earliest_frames=True)
+                    num_updates = config.num_updates_per_train_iter
+                else:
+                    assert config.mini_batch_length is not None, (
+                        "No mini_batch_length is specified for off-policy training"
+                    )
+                    experience, batch_info = self._replay_buffer.get_batch(
+                        batch_size=(mini_batch_size *
+                                    config.num_updates_per_train_iter),
+                        batch_length=config.mini_batch_length)
+                    num_updates = 1
+            return experience, batch_info, num_updates, mini_batch_size
+
+        if not self.has_offline:
+            experience, batch_info, num_updates, mini_batch_size = _replay()
+            with record_time("time/train"):
+                return self._train_experience(
+                    experience,
+                    batch_info,
+                    num_updates,
+                    mini_batch_size,
+                    config.mini_batch_length,
+                    (config.update_counter_every_mini_batch
+                     and update_global_counter),
+                    whole_replay_buffer_training=config.
+                    whole_replay_buffer_training)
+        else:
+            # hybrid training scheme
+            global_step = alf.summary.get_global_counter()
+            if ((global_step >= self._rl_train_after_update_steps) and
+                (global_step % self._rl_train_every_update_steps == 0)):
+                self._RL_train = True
+            else:
+                self._RL_train = False
+
+            if global_step >= self._rl_train_after_update_steps:
+                # flag used for indicating initial pre-training phase
+                self._pre_train = False
             else:
-                assert config.mini_batch_length is not None, (
-                    "No mini_batch_length is specified for off-policy training"
+                self._pre_train = True
+
+            if self._RL_train:
+                experience, batch_info, num_updates, mini_batch_size = _replay(
                 )
-                experience, batch_info = self._replay_buffer.get_batch(
+            else:
+                experience = None
+                batch_info = None
+                num_updates = 1
+                mini_batch_size = config.mini_batch_size
+
+            with record_time("time/replay"):
```

Comment:
`"time/offline_replay" to differentiate with "time/replay"

Reply:
Changed